src/Ema/Route/Generic/Verification.hs

```diff
@@ -0,0 +1,163 @@
+          P.% ty)))
+      (TypeError 
+        (model 
+        P.% "Is not a supermodel of the model:" 
+        P.% ty
+        P.% "Even though this is required by its WithSubRoutes list."))
+
+type family RouteShapeMismatchError r where
+  RouteShapeMismatchError r =  TypeError ("WithSubRoutes list does not match the shape of " P.<> r)
+
+-- Attempts to 'unwrap' @r2@ to see if the constructor fields specified by @r1@ match its internal representation 1:1
+type family IsUnwrappedRoute (r1 :: [Type]) (r2 :: Type) :: Bool where
+  -- For routes that derived /stock/ GHC.Generic;
+  -- TODO: The implementation is a bit overkill here as it checks for all fields, but this could be useful
+  -- should semantics expand in the future perhaps?
+  IsUnwrappedRoute ts (GHC.D1 _ (GHC.C1 _ fields) _) = 
+    IsUnwrappedRoute ts (fields ())
+  IsUnwrappedRoute (t ': '[]) (GHC.S1 _ (GHC.K1 _ t') _) =
+    t == t'
+  IsUnwrappedRoute (t ': ts) ((GHC.S1 _ (GHC.K1 _ t') GHC.:*: nxt) _) =
+    t == t' && IsUnwrappedRoute ts (nxt ())
+  -- Special case for routes with no fields internally, since we can think of Unwrapped () ~ ()
+  IsUnwrappedRoute '[] (GHC.U1 ()) = 
+    'True
+  -- Catch-all
+  IsUnwrappedRoute _ _ = 
+    'False
+
+-- We need to implement the matching logic as 2 type families here due to overlapping patterns
+type family IsUnwrappedRoute' (r1 :: [Type]) (r2 :: Type) :: Bool where
+  -- For routes that derived /newtype/ GHC.Generic; simply verify the reps are equal
+  -- Otherwise, pass it on to match with the assumption of /stock/ GHC.Generic deriving
+  IsUnwrappedRoute' (t ': ts) ts' = 
+    GHC.Rep t () == ts' || IsUnwrappedRoute (t ': ts) ts'
+  IsUnwrappedRoute' r1 r2 = 
+    IsUnwrappedRoute r1 r2
+
+{- | @VerifyRoutes route rep subroutes@ verifies the given @route@ to ensure that there
+exists a valid @HasSubRoutes@ instance for @route@ given its @rep@ and the @subroutes@ it is generic-deriving from. 
+
+Invariant: rep ~ Code route
+-}
+type family VerifyRoutes (route :: Type) (rep :: [[Type]]) (subroutes :: [Type]) :: Constraint where
+  VerifyRoutes _ '[] '[] = ()
+  -- Inconsistent lengths
+  VerifyRoutes r '[]  _  = RouteShapeMismatchError r
+  VerifyRoutes r _   '[] = RouteShapeMismatchError r
+  -- Subroute rep is unit
+  VerifyRoutes r ('[] ': rs) (() : rs') = VerifyRoutes r rs rs'
+  VerifyRoutes r ('[()] ': rs) (() : rs') = VerifyRoutes r rs rs'
+  VerifyRoutes r (r' ': rs) (() : rs') = 
+    TypeError 
+      ("WithSubRoutes list states that the route constructor at this index should only contain () or be empty"
+      P.% "But it is " P.<> r')
+  -- Constructor type ~ Subroute spec
+  VerifyRoutes r ('[r'] ': rs) (r' : rs') = VerifyRoutes r rs rs'
+  -- Constructor type ~ Unwrapped (Subroute spec) as a last-resort assumption
+  -- TODO: Type specified may not be GHC.Generic; might be better to dispatch via class instance.
+  VerifyRoutes r (r1 ': rs) (r2 ': rs') = 
+    If (r1 `IsUnwrappedRoute'` (GHC.Rep r2 ()))
```

Comment:
I am yet closely read this, so I might be wrong, but shouldn't we merely be checking that `GIsomorphic` constraint exists instead of all these unwrapping checks?

For any sub-route `r`, you are allowed to use *any* type `t` in the corresponding index of `WithSubRoutes` as long as the `GIsomorphic r t` constraint is satisfied.

Reply:
We could also go back to using `Coercible` instead of `GIsomorphic` FWIW.

Reply:
I've ran into a bit of a problem with this:

```hs
-- | Dispatched verification to check if 'GIsomorphic' can be satisfied, otherwise instance matching
-- will fall on a type error.
class IsUnwrappedRoute (r1 :: [Type]) (r2 :: Type) where

instance {-# OVERLAPPING #-} GIsomorphic r1 r2 => IsUnwrappedRoute (r1 ': rs) r2 where

instance {-# OVERLAPPABLE #-}
  ( TypeError
      ( "Route constructor with representation:"
          P.% r1
          P.% "Does not contain a type matching the subroute, or an unwrapped representation of the subroute:"
          P.% r2
          P.% "As specified in its (potentially inferred) WithSubRoutes list."
      )
  ) => IsUnwrappedRoute r1 r2 where
```
Normally this is non-problematic, but since we have to do the `~ (() :: Constraint` hack to get deriving clauses to infer the context correctly, this does not unify. We end up with `IsUnwrappedRoute` constraints that cannot be reduced further, causing this error:

```
    • Couldn't match type ‘(Ema.Route.Generic.Verification.IsUnwrappedRoute
                              ('[] @Type) (FileRoute "index.html"),
                            (Ema.Route.Generic.Verification.IsUnwrappedRoute
                               ('[] @Type) (FileRoute "onlytime.html"),
                             () :: Constraint))’
                     with ‘() :: Constraint’
```

Reply:
Hmm, I really wish we could use `GIsomorphic` (or `Coercible` in some way), but apparently Haskell is not there yet.